### PR TITLE
[processor/ratelimit] Add new component ratelimitprocessor

### PR DESCRIPTION
**Pull request by carsonip**
Add new processor for configurable rate limiting. Use Apache 2.0 license.

### CODE REVIEW

File: processor/ratelimitprocessor/internal/sharedcomponent/sharedcomponent.go [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// Package sharedcomponent exposes functionality for components
+// to register against a shared key, such as a configuration object, in order to be reused across signal types.
+// This is particularly useful when the component relies on a shared resource such as os.File or http.Server.
+package sharedcomponent // import "github.com/elastic/opentelemetry-collector-components/processor/ratelimitprocessor/internal/sharedcomponent"
```
Comment (vigneshshanmugam) [inline processor/ratelimitprocessor/internal/sharedcomponent/sharedcomponent.go:21]: WDYT about moving this as separate package given we are also using in elasticapmreceiver - https://github.com/elastic/opentelemetry-collector-components/tree/main/receiver/elasticapmreceiver/internal/sharedcomponent ?

**Review by vigneshshanmugam — APPROVED** [review]
LGTM

File: processor/ratelimitprocessor/internal/sharedcomponent/sharedcomponent.go (hunk 1 above)
Comment (vigneshshanmugam) [reply processor/ratelimitprocessor/internal/sharedcomponent/sharedcomponent.go:27]: Thanks for the details, I cant think of a way from top of my head. Its not a blocker per se. I am fine with keeping it as it is.

**Review by axw — APPROVED** [review]
thanks!

**Comment by axw** [comment]
@constanca-m ICYMI

**Pull request merged**